packages/insured-bridge-relayer/src/index.ts

```diff
@@ -51,129 +55,155 @@ export async function run(logger: winston.Logger, l1Web3: Web3): Promise<void> {
-      l2Client,
-      profitabilityCalculator,
-      filteredL1Whitelist,
-      accounts[0],
-      config.whitelistedChainIds,
-      config.l1DeployData,
-      config.l2DeployData,
-      config.l2BlockLookback
-    );
-
-    const canonicalBridgeAdapter = createBridgeAdapter(logger, l1Web3, l2Web3, config.activatedChainIds[0]);
-    if (config.botModes.l1FinalizerEnabled) await canonicalBridgeAdapter.initialize();
-
-    const crossDomainFinalizer = new CrossDomainFinalizer(
-      logger,
-      gasEstimator,
-      l1Client,
-      l2Client,
-      canonicalBridgeAdapter,
-      accounts[0],
-      config.l2DeployData,
-      config.crossDomainFinalizationThreshold
-    );
-
+    const combinedFilteredL1Whitelist = lodash.uniq(relayers.map((relayer) => relayer.filteredL1Whitelist).flat());
+    await approveL1Tokens(logger, l1Web3, gasEstimator, accounts[0], config.bridgeAdmin, combinedFilteredL1Whitelist);
     for (;;) {
-      await retry(
-        async () => {
-          // Update state.
-          await Promise.all([
-            gasEstimator.update(),
-            l1Client.update(),
-            l2Client.update(),
-            profitabilityCalculator.update(),
-          ]);
-
-          // Start bots that are enabled.
-          if (config.botModes.relayerEnabled) await relayer.checkForPendingDepositsAndRelay();
-          else logger.debug({ at: "AcrossRelayer#Relayer", message: "Relayer disabled" });
-
-          if (config.botModes.disputerEnabled) await relayer.checkForPendingRelaysAndDispute();
-          else logger.debug({ at: "AcrossRelayer#Disputer", message: "Disputer disabled" });
-
-          if (config.botModes.settlerEnabled) await relayer.checkforSettleableRelaysAndSettle();
-          else logger.debug({ at: "AcrossRelayer#Finalizer", message: "Relay Settler disabled" });
-
-          if (config.botModes.l1FinalizerEnabled) await crossDomainFinalizer.checkForConfirmedL2ToL1RelaysAndFinalize();
-          else logger.debug({ at: "AcrossRelayer#CrossDomainFinalizer", message: "Confirmed L1 finalizer disabled" });
-
-          if (config.botModes.l2FinalizerEnabled) await crossDomainFinalizer.checkForBridgeableL2TokensAndBridge();
-          else logger.debug({ at: "AcrossRelayer#CrossDomainFinalizer", message: "L2->L1 finalizer disabled" });
-
-          // Each of the above code blocks could have produced transactions. If they did, their promises are stored
-          // in the executed transactions array. The method below awaits all these transactions to ensure they are
-          // correctly included in a block. if any submitted transactions contains an error then a log is produced.
-          await processTransactionPromiseBatch(
-            [...relayer.getExecutedTransactions(), ...crossDomainFinalizer.getExecutedTransactions()],
-            logger
+      const outputs = await Promise.allSettled(
+        relayers.map(async ({ relayer, crossDomainFinalizer, l2Client, profitabilityCalculator }) => {
```

Comment:
Similar to the above, this just moves all of the logic in here into a Promise.allSettled loop over the return values from above. The array just stores all the L2-specific objects that we'll need for this loop. Otherwise, all the logic is exactly the same as before.

Note: allSettled is used to ensure that one failing L2 doesn't stop the others from getting through their runs. Only once all runs have been completed do we throw an error.